### PR TITLE
fix: use new vault notifiers

### PR DESCRIPTION
**Pull request by samsiegart**
fixes: https://github.com/Agoric/agoric-sdk/issues/4882

A few other fixes included:
- Only show errored and "closed" vaults when "show closed vaults" is checked, it used to show all vaults
- Show the "show closed vaults" checkbox if there's an errored vault but no "closed" vaults. It used to only let you see errored vaults if there was also at least one "closed" vault.
- Fix the collateralization ratio calculation in the vault summary cards & vault management page, it broke when using denoms other than BLD and RUN (different decimal places).

### CODE REVIEW

File: ui/src/components/Treasury.jsx
Comment (turadg) [inline ui/src/components/Treasury.jsx:125]: this filter expression seems worth pulling out into a function.

how many different status values are there? could they be enumerated? if so the filter function could be O(1) by looking up in a table. Not perf critical or a blocker but something to consider.

Either way these strings seem easy to mistype so I suggest either a function or constants.

File: ui/src/contexts/Application.jsx
Comment (turadg) [inline ui/src/contexts/Application.jsx:73]: without the asset notifier the UI can't show the actual debt. It has to be calculated from the vault's debtSnapshot and the manager/asset's current compounded interest. Like so https://github.com/Agoric/agoric-sdk/blob/52f60eb/packages/run-protocol/src/vaultFactory/vault.js#L218-L223

Is that in scope for this PR?

File: ui/src/components/Treasury.jsx
Comment (samsiegart) [reply ui/src/components/Treasury.jsx]: +1 on constants, just made that change here and throughout the dapp.

I think the filter function is pretty self-documenting and only used once, it's literally just figuring out which vaults should be rendered. So I'm not too eager to make it its own function, but open to change that.

I'm not sure exactly what you mean by the O(1) table lookup, would you mind spelling out an example? We need to preprocess this list to know its length for the rendering logic below, which would be O(n). Unless you mean the `Array.includes` thing should be `Set.has` or something, but I think `Array.includes` is actually pretty fast on two elements:
jsbench

File: ui/src/components/VaultSummary.jsx
Comment (samsiegart) [reply ui/src/components/VaultSummary.jsx]: Good point, turns out it's not needed, I just based it off some other example I saw that was hardening

File: ui/src/components/VaultSummary.jsx
Comment (samsiegart) [reply ui/src/components/VaultSummary.jsx]: Done thanks.

I'm not too eager to jump to TypeScript. It's probably more likely to pay off the more complex a dapp becomes, but not sure if it's worth rewriting this one at this point. It's sometimes too tempting to slap a bunch of `any`s on things especially when dealing with external libraries or browser APIs, so it requires a certain amount of commitment to get value out of it. I could probably be talked into it with some pertinent examples, but I don't feel as though types have been a limiting factor in productivity so far.

It would definitely be good to have some example dapp template that uses TypeScript, but also important that we maintain some templates using JavaScript.

File: ui/src/components/VaultSummary.jsx
Comment (samsiegart) [reply ui/src/components/VaultSummary.jsx:101]: Related to your above comment, vault is type `VaultData` [ui/src/types/types.js](https://github.com/Agoric/dapp-treasury/pull/94/files/1fdcce49c6f8b9d4d62327f9ab2cdd7bbdcc272a#diff-eb2331b2e81f8a38629f121d4974ca0d4c7844e4c61a481d853f980f3c7b9a42) which is slightly different from `VaultUIState`. It would be elegant to have these be the same types, but I'm unsure if we should try to make them conform.

For example, the `status` is partially informed by the `VaultPhase`, but it can also be something like `Error in Offer`, `Loading`, or `Pending wallet approval`. These mean different things to the UI, but aren't part of the contract. There's also the `err` property.

In Web2 applications I've worked on, it's been standard practice for the UI to have its own types, since inevitably there's always transformations that need to be done to backend data to make it convenient/usable to the UI code. I guess with OCAPS this line is a bit blurred, and the transformations are minimal in some places, but I think it will be easier to not paint ourselves into a corner if we maintain this separation.

File: ui/src/contexts/Application.jsx
Comment (samsiegart) [reply ui/src/contexts/Application.jsx:74]: I actually wasn't sure how that was supposed to work, thanks. I think it's worth merging this first since that seems like a separate step, and the dapp crashes without this current PR as is. I can create a separate issue for that.

File: ui/src/components/Treasury.jsx
Comment (turadg) [reply ui/src/components/Treasury.jsx]: > I think the filter function is pretty self-documenting and only used once, it's literally just figuring out which vaults should be rendered. So I'm not too eager to make it its own function, but open to change that.

SGTM.

> I'm not sure exactly what you mean by the O(1) table lookup, would you mind spelling out an example?

Screen Shot 2022-04-01 at 7 54 31 AM


> We need to preprocess this list to know its length for the rendering logic below, which would be O(n).

I meant the runtime of the function being passed to filter.

> Array.includes is actually pretty fast on two elements:

Definitely. The speed gain from the object isn't worth the typing. The value is more in naming.

```
  const vaultsToRender = vaultsList.filter(([_, {status}]) =>
    showClosed ? renderAlways[status] : renderOpen[status]
```

Again, a "consider" not a requested change.

File: ui/src/components/VaultSummary.jsx
Comment (turadg) [reply ui/src/components/VaultSummary.jsx]: That's cool. Though there wouldn't be any rewriting. We'd just change the extension of some files to `.ts` and then be able to use that syntax. We aren't doing that in most of SDK because it would entail a build step, but JSX already requires a build step.

**Review by turadg — COMMENTED** [review]
Some more comments. I'll approve when the vault status strings aren't duplicated.

File: ui/src/constants.js
Comment (samsiegart) [reply ui/src/constants.js]: Done

File: ui/src/components/Treasury.jsx
Comment (samsiegart) [reply ui/src/components/Treasury.jsx]: Cleaned this function up a bit as discussed

File: ui/src/components/VaultSummary.jsx
Comment (samsiegart) [reply ui/src/components/VaultSummary.jsx]: Ack. As discussed I think we want to look into typescript at some point soon

File: ui/src/types/types.js
Comment (samsiegart) [reply ui/src/types/types.js]: Changed as discussed to use more of the types from `run-protocol` where appropriate

File: ui/src/components/vault/VaultManagement/VaultManagement.jsx
Comment (turadg) [inline ui/src/components/vault/VaultManagement/VaultManagement.jsx:62]: nice

File: ui/src/components/VaultSummary.jsx
Comment (samsiegart) [reply ui/src/components/VaultSummary.jsx:101]: Changed as discussed to use more of the types from run-protocol where appropriate

**Comment by samsiegart** [comment]
I had to undo some of the changes we made with the `Partial<VaultData>` stuff because we actually need undefined for errored/pending vaults etc. which only have an id and status.

File: ui/src/components/vault/VaultManagement/VaultManagement.jsx
Comment (turadg) [inline ui/src/components/vault/VaultManagement/VaultManagement.jsx:72]: now that these values come from a full `VaultData`, I bet this condition never fails and could be removed.

File: ui/src/components/vault/VaultManagement/VaultManagement.jsx
Comment (turadg) [inline ui/src/components/vault/VaultManagement/VaultManagement.jsx:75]: pretty sure this is redundant too since if either of these were falsy that line 73 would leave this function

**Review by turadg — APPROVED** [review]
Thanks for the collateral improvements

File: ui/src/components/vault/VaultManagement/VaultManagement.jsx
Comment (samsiegart) [reply ui/src/components/vault/VaultManagement/VaultManagement.jsx]: You're right. The only way these could be undefined is if we have a pending/closed/errored vault, which we don't let the user navigate to this "management" view for.

File: ui/src/components/vault/VaultManagement/VaultManagement.jsx
Comment (samsiegart) [reply ui/src/components/vault/VaultManagement/VaultManagement.jsx:72]: So I needed to add back in the undefined for reasons stated in my other comment, making these asserts necessary, but I was at least able to remove the `null`s from the `VaultData` type now

**Pull request merged**
